### ISSUE TITLE
Downloading eslint,vue-language-server...

### ISSUE THREAD

**Issue by lq0910** [zed-industries/zed#18258]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

一直下载中.... 语言服务无法使用 ，代码提示和函数方法跳转也不起作用

### Environment

Zed: v0.155.0 (Zed Dev 3ca18af40b8a7cb83d8303a8131e90ca997f09ca)
OS: Windows 10.0.26100
Memory: 15.7 GiB
Architecture: x86_64
GPU: Intel(R) Iris(R) Xe Graphics || Intel Corporation || Intel driver

### If applicable, add mockups / screenshots to help explain present your vision of the feature

![image](https://github.com/user-attachments/assets/14cebb0d-0904-4771-bfc8-e588818a9eeb)
![image](https://github.com/user-attachments/assets/e25d66fd-fb34-4597-aad4-3ccef2c37539)
![image](https://github.com/user-attachments/assets/3485b993-49b3-4d77-8919-adbdb3f84c79)

### If applicable, attach your Zed.log file to this issue.

Caused by:
    0: request or operation took longer than the configured timeout time
    1: [28] Timeout was reached
2024-09-24T12:54:33.2866712+08:00 [ERROR] failed to start language server "eslint": failed to launch npm subcommand install subcommand
err: Some(error downloading Node binary tarball

**Comment by lq0910** [comment]
Same issue with my macbook

**Comment by bestgopher** [comment]
It might be due to network issues. You could set up a proxy like this:

![image](https://github.com/user-attachments/assets/c08942bc-d702-49c5-a416-42ed1e361013)

**Comment by lq0910** [comment]
已收到。我会马上查看。

**Comment by lq0910** [comment]
@bestgopher  Thank you very much. I have fixed the version below 0.153 and there is no need to configure a proxy. But Vue's language server has encountered another error

**Comment by lq0910** [comment]
```2024-09-25T09:04:46.2754747+08:00 [WARN] test binary check failed
2024-09-25T09:04:46.2755109+08:00 [INFO] beginning to reinstall server
2024-09-25T09:04:46.275561+08:00 [INFO] deleting server container
2024-09-25T09:04:46.275932+08:00 [ERROR] server container removal

Caused by:
    系统找不到指定的文件。 (os error 2)
2024-09-25T09:04:46.2759981+08:00 [INFO] starting language server "vue-language-server", path: "D:\\docs\\project\\zq-slj\\zq-slj-app-frontend\\sz_h5", id: 19
2024-09-25T09:04:46.2777072+08:00 [INFO] Node runtime install_if_needed
2024-09-25T09:04:46.4151099+08:00 [INFO] starting language server. binary path: "C:\\Users\\liqiang\\AppData\\Local\\Zed\\node\\node-v22.5.1-win-x64\\node.exe", working directory: "D:\\docs\\project\\zq-slj\\zq-slj-app-frontend\\sz_h5", args: ["/C:\\Users\\liqiang\\AppData\\Local\\Zed\\extensions\\work\\vue/node_modules/@vue/language-server/bin/vue-language-server.js", "--stdio"]
2024-09-25T09:04:46.4454793+08:00 [INFO] Node runtime install_if_needed
2024-09-25T09:04:46.4682538+08:00 [ERROR] cannot read LSP message headers
2024-09-25T09:04:46.4684361+08:00 [ERROR] failed to start language server "vue-language-server": oneshot canceled
2024-09-25T09:04:46.4684911+08:00 [ERROR] server stderr: Some("node:internal/modules/cjs/loader:1222\r\n  throw err;\r\n  ^\r\n\r\nError: Cannot find module 'D:\\C:\\Users\\liqiang\\AppData\\Local\\Zed\\extensions\\work\\vue\\node_modules\\@vue\\language-server\\bin\\vue-language-server.js'\r\n    at Module._resolveFilename (node:internal/modules/cjs/loader:1219:15)\r\n    at Module._load (node:internal/modules/cjs/loader:1045:27)\r\n    at TracingChannel.traceSync (node:diagnostics_channel:315:14)\r\n    at wrapModuleLoad (node:internal/modules/cjs/loader:215:24)\r\n    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:158:5)\r\n    at node:internal/main/run_main_module:30:49 {\r\n  code: 'MODULE_NOT_FOUND',\r\n  requireStack: []\r\n}\r\n\r\nNode.js v22.5.1\r\n")
2024-09-25T09:04:46.4686643+08:00 [INFO] retrying installation of language server "vue-language-server" in 1s
2024-09-25T09:04:46.4687722+08:00 [ERROR] server shut down
2024-09-25T09:04:46.5705457+08:00 [INFO] Node runtime install_if_needed
2024-09-25T09:04:47.469016+08:00 [INFO] About to spawn test binary
2024-09-25T09:04:47.4691925+08:00 [WARN] test binary failed to launch
2024-09-25T09:04:47.4693048+08:00 [WARN] test binary check failed
2024-09-25T09:04:47.4694103+08:00 [INFO] beginning to reinstall server
2024-09-25T09:04:47.4695305+08:00 [INFO] deleting server container
2024-09-25T09:04:47.4703309+08:00 [ERROR] server container removal

Caused by:
    系统找不到指定的文件。 (os error 2)
2024-09-25T09:04:47.4704863+08:00 [INFO] starting language server "vue-language-server", path: "D:\\docs\\project\\zq-slj\\zq-slj-app-frontend\\sz_h5", id: 20
2024-09-25T09:04:47.4727081+08:00 [INFO] Node runtime install_if_needed
2024-09-25T09:04:47.5956061+08:00 [INFO] starting language server. binary path: "C:\\Users\\liqiang\\AppData\\Local\\Zed\\node\\node-v22.5.1-win-x64\\node.exe", working directory: "D:\\docs\\project\\zq-slj\\zq-slj-app-frontend\\sz_h5", args: ["/C:\\Users\\liqiang\\AppData\\Local\\Zed\\extensions\\work\\vue/node_modules/@vue/language-server/bin/vue-language-server.js", "--stdio"]
2024-09-25T09:04:47.645386+08:00 [ERROR] cannot read LSP message headers
2024-09-25T09:04:47.6455988+08:00 [ERROR] failed to start language server "vue-language-server": oneshot canceled
2024-09-25T09:04:47.6457459+08:00 [ERROR] server stderr: Some("node:internal/modules/cjs/loader:1222\r\n  throw err;\r\n  ^\r\n\r\nError: Cannot find module 'D:\\C:\\Users\\liqiang\\AppData\\Local\\Zed\\extensions\\work\\vue\\node_modules\\@vue\\language-server\\bin\\vue-language-server.js'\r\n    at Module._resolveFilename (node:internal/modules/cjs/loader:1219:15)\r\n    at Module._load (node:internal/modules/cjs/loader:1045:27)\r\n    at TracingChannel.traceSync (node:diagnostics_channel:315:14)\r\n    at wrapModuleLoad (node:internal/modules/cjs/loader:215:24)\r\n    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:158:5)\r\n    at node:internal/main/run_main_module:30:49 {\r\n  code: 'MODULE_NOT_FOUND',\r\n  requireStack: []\r\n}\r\n\r\nNode.js v22.5.1\r\n")
2024-09-25T09:04:47.6459154+08:00 [ERROR] Hit 4 reinstallation attempts for "vue-language-server"
2024-09-25T09:04:47.645976+08:00 [ERROR] server shut down
2024-09-25T09:04:48.3037061+08:00 [INFO] downloading language server "vtsls"
2024-09-25T09:04:48.313555+08:00 [INFO] Node runtime install_if_needed
2024-09-25T09:04:48.4367243+08:00 [INFO] starting language server. binary path: "C:\\Users\\liqiang\\AppData\\Local\\Zed\\node\\node-v22.5.1-win-x64\\node.exe", working directory: "D:\\docs\\project\\zq-slj\\zq-slj-app-frontend\\sz_h5", args: ["C:\\Users\\liqiang\\AppData\\Local\\Zed\\languages\\vtsls\\node_modules/@vtsls/language-server/bin/vtsls.js", "--stdio"]
2024-09-25T09:04:49.7413722+08:00 [INFO] Creating buffer 0x15343f2bc58 of size 64288, name 'chunk-11', handle 14
2024-09-25T09:04:49.7449175+08:00 [INFO] Creating buffer 0x15343f2aba8 of size 64288, name 'chunk-12', handle 15
2024-09-25T09:04:49.9897651+08:00 [INFO] Creating buffer 0x15343f2ddb8 of size 75712, name 'chunk-13', handle 16
2024-09-25T09:04:50.0015633+08:00 [INFO] Creating buffer 0x15343f2b138 of size 89600, name 'chunk-14', handle 17
2024-09-25T09:04:50.0290107+08:00 [INFO] Creating buffer 0x15343f2d298 of size 99680, name 'chunk-15', handle 18
2024-09-25T09:04:50.0547858+08:00 [INFO] Creating buffer 0x15343f2cd08 of size 100128, name 'chunk-16', handle 19
2024-09-25T09:04:50.3385536+08:00 [INFO] Creating buffer 0x15343f2a088 of size 101584, name 'chunk-17', handle 20
2024-09-25T09:04:50.376602+08:00 [INFO] Creating buffer 0x15343f2d828 of size 113344, name 'chunk-18', handle 21
2024-09-25T09:04:52.2828739+08:00 [INFO] Creating buffer 0x15343f2a618 of size 113344, name 'chunk-19', handle 22
2024-09-25T09:05:00.4261958+08:00 [INFO] Creating texture 0x153c0223538 of size 1024x1024x1 and format R16Float, name 'atlas', handle 23
2024-09-25T09:05:00.4263614+08:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-09-25T09:05:00.4310532+08:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-09-25T09:05:00.4425903+08:00 [INFO] Creating a descriptor pool for at most 256 sets
2024-09-25T09:05:00.4563819+08:00 [INFO] Creating a descriptor pool for at most 256 sets
2024-09-25T09:05:05.7446636+08:00 [INFO] Creating texture 0x153c0223fc8 of size 1030x1024x1 and format R16Float, name 'atlas', handle 24```

**Comment by bestgopher** [comment]
> @bestgopher Thank you very much. I have fixed the version below 0.153 and there is no need to configure a proxy. But Vue's language server has encountered another error

I'm not familiar with Vue, so I can't offer any suggestions.

**Comment by lq0910** [comment]
@bestgopher  okay, thank you

**Comment by JunkuiZhang** [comment]
Any extension you download from the extension store will not work for now. See #15004

**Comment by lq0910** [comment]
@JunkuiZhang  我工作主要是vue，目前其他扩展基本可以使用，只有vue无法使用，html的语法提示目前也不起作用，就没有什么办法解决？

**Comment by JunkuiZhang** [comment]
@lq0910 你会改代码吗？
```bash
diff --git a/crates/extension/src/extension_lsp_adapter.rs b/crates/extension/src/extension_lsp_adapter.rs
index 25179acec..ccc1cbf5e 100644
--- a/crates/extension/src/extension_lsp_adapter.rs
+++ b/crates/extension/src/extension_lsp_adapter.rs
@@ -88,7 +88,17 @@ impl LspAdapter for ExtensionLspAdapter {

             Ok(LanguageServerBinary {
                 path,
-                arguments: command.args.into_iter().map(|arg| arg.into()).collect(),
+                arguments: command
+                    .args
+                    .into_iter()
+                    .map(|arg| {
+                        if let Some(arg) = arg.strip_prefix('/') {
+                            arg.into()
+                        } else {
+                            arg.into()
+                        }
+                    })
+                    .collect(),
                 env: Some(command.env.into_iter().collect()),
             })
         }
```

**Comment by lq0910** [comment]
可以改的，只是我现在本地编译因为网络原有总是编译失败

**Comment by lq0910** [comment]
@JunkuiZhang 这段代码我看和#15004 里面的不太一样哦

**Comment by JunkuiZhang** [comment]
解决的是不一样的问题

**Comment by lq0910** [comment]
好的，谢谢。 我改了试试，我问了下别人扩展都可以用win版本，真是怪了

**Comment by lq0910** [comment]
你用的是自己编译的版本？

**Comment by JunkuiZhang** [comment]
> 你用的是自己编译的版本？

是的，尽量不要用别人编译好的二进制文件，安全考虑

**Comment by lq0910** [comment]
@JunkuiZhang  好的，非常感谢！

**Comment by provenceo** [comment]
FinalizationRegistry is not defined
<img width="724" alt="image" src="https://github.com/user-attachments/assets/3b1f9a37-3e9b-406b-997f-23508d071a1c">

**Comment by SomeoneToIgnore** [comment]
I had a look and indeed, I can reliably reproduce this:
open a bunch of various files (so there's multiple LSP servers to download to be sure to hit the issue)
start a new zed with internet on
instantly turn the internet off when the download starts
:boom: one or more servers are stuck in the status bar
Here's what I see in the logs (note the timings):

```
[2024-09-24T23:38:08+03:00 INFO  language::language_registry] attempting to start language server "yaml-language-server", path: "/Users/someonetoignore/work/other/local_test", id: 1
[2024-09-24T23:38:10+03:00 INFO  language] fetching latest version of language server "yaml-language-server"
[2024-09-24T23:38:18+03:00 INFO  language] downloading language server "yaml-language-server"
[2024-09-24T23:39:57+03:00 ERROR project::lsp_store] failed to start language server "yaml-language-server": failed to execute npm install subcommand:
stdout: ""
stderr: "npm error code ENOTFOUND\nnpm error syscall getaddrinfo\nnpm error errno ENOTFOUND\nnpm error network request to https://registry.npmjs.org/ajv failed, reason: getaddrinfo ENOTFOUND registry.npmjs.org\nnpm error network This is a problem related to network connectivity.\nnpm error network In most cases you are behind a proxy or have bad network settings.\nnpm error network\nnpm error network If you are behind a proxy, please make sure that the\nnpm error network 'proxy' config is set properly.  See: 'npm help config'\nnpm error A complete log of this run can be found in: /Users/someonetoignore/Library/Application Support/Zed/node/cache/_logs/2024-09-24T20_38_18_795Z-debug-0.log\n"
[2024-09-24T23:39:57+03:00 ERROR project::lsp_store] server stderr: Some("")
[2024-09-24T23:39:57+03:00 INFO  project::lsp_store] retrying installation of language server "yaml-language-server" in 1s
[2024-09-24T23:39:58+03:00 ERROR languages] no cached binary

Stack backtrace:
....snip
[2024-09-24T23:39:58+03:00 INFO  project::lsp_store] About to spawn test binary
[2024-09-24T23:39:58+03:00 WARN  project::lsp_store] test binary failed to launch
[2024-09-24T23:39:58+03:00 WARN  project::lsp_store] test binary check failed
[2024-09-24T23:39:58+03:00 INFO  project::lsp_store] beginning to reinstall server
[2024-09-24T23:39:58+03:00 INFO  language::language_registry] deleting server container
[2024-09-24T23:39:58+03:00 INFO  language::language_registry] attempting to start language server "yaml-language-server", path: "/Users/someonetoignore/work/other/local_test", id: 9
[2024-09-24T23:39:58+03:00 INFO  language] fetching latest version of language server "yaml-language-server"
[2024-09-24T23:40:05+03:00 INFO  language] downloading language server "yaml-language-server"
[2024-09-24T23:41:44+03:00 ERROR languages] no cached binary

etc. etc. etc.
```

Notice how long it takes to get from
```
[2024-09-24T23:38:18+03:00 INFO  language] downloading language server "yaml-language-server"
```
'till
```
[2024-09-24T23:39:57+03:00 ERROR project::lsp_store] failed to start language server "yaml-language-server": failed to execute npm install subcommand:
```
(this is just one attempt)
This is due to 2 npm network access attempts:
```rs
let should_install_language_server = self
    .node
    .should_install_npm_package(package_name, &server_path, &container_dir, &latest_version)
    .await;

if should_install_language_server {
    self.node
        .npm_install_packages(&container_dir, &[(package_name, latest_version.as_str())])
        .await?;
}
```
both having this:

```
"--fetch-retry-mintimeout",
"2000",
"--fetch-retry-maxtimeout",
"5000",
"--fetch-timeout",
"5000",```

The units are supposed to be milliseconds, according to https://docs.npmjs.com/cli/v9/using-npm/config#fetch-retries ; and we will do the attempt twice according to the docs:
```
fetch-retries

    Default: 2
    Type: Number

The "retries" config for the retry module to use when fetching packages from the registry.

npm will retry idempotent read requests to the registry in the case of network failures or 5xx HTTP errors.
```

Both commands are supposed to use this parameter, but apparently it's not about accessing the registry but about downloading the packages only!
Because this is what I see in the npm logs (full logs attached):

```
16 silly packumentCache full:https://registry.npmjs.org/yaml-language-server cache-miss
17 http fetch GET https://registry.npmjs.org/yaml-language-server attempt 1 failed with ENOTFOUND
18 http fetch GET https://registry.npmjs.org/yaml-language-server attempt 2 failed with ENOTFOUND
19 http fetch GET https://registry.npmjs.org/yaml-language-server attempt 3 failed with ENOTFOUND
20 http fetch GET 200 https://registry.npmjs.org/yaml-language-server 7069ms (cache stale)
21 silly packumentCache full:https://registry.npmjs.org/yaml-language-server set size:6448684 disposed:false
22 silly placeDep ROOT yaml-language-server@1.15.0 OK for:  want: 1.15.0
23 silly fetch manifest ajv@^8.11.0
24 silly packumentCache full:https://registry.npmjs.org/ajv cache-miss
25 silly fetch manifest lodash@4.17.21
26 silly packumentCache full:https://registry.npmjs.org/lodash cache-miss
27 silly fetch manifest request-light@^0.5.7
28 silly packumentCache full:https://registry.npmjs.org/request-light cache-miss
29 silly fetch manifest vscode-json-languageservice@4.1.8
30 silly packumentCache full:https://registry.npmjs.org/vscode-json-languageservice cache-miss
31 silly fetch manifest vscode-languageserver@^7.0.0
32 silly packumentCache full:https://registry.npmjs.org/vscode-languageserver cache-miss
33 silly fetch manifest vscode-languageserver-textdocument@^1.0.1
34 silly packumentCache full:https://registry.npmjs.org/vscode-languageserver-textdocument cache-miss
35 silly fetch manifest vscode-languageserver-types@^3.16.0
36 silly packumentCache full:https://registry.npmjs.org/vscode-languageserver-types cache-miss
37 silly fetch manifest vscode-nls@^5.0.0
38 silly packumentCache full:https://registry.npmjs.org/vscode-nls cache-miss
39 silly fetch manifest vscode-uri@^3.0.2
40 silly packumentCache full:https://registry.npmjs.org/vscode-uri cache-miss
41 http fetch GET https://registry.npmjs.org/lodash attempt 1 failed with ENOTFOUND
42 http fetch GET https://registry.npmjs.org/request-light attempt 1 failed with ENOTFOUND
43 http fetch GET https://registry.npmjs.org/ajv attempt 1 failed with ENOTFOUND
44 http fetch GET https://registry.npmjs.org/vscode-languageserver attempt 1 failed with ENOTFOUND
45 http fetch GET https://registry.npmjs.org/vscode-json-languageservice attempt 1 failed with ENOTFOUND
...
```

So, not only registry checks ignore those parameters, they seem to do 3 retries per package.

---------------

TL; DR
The stuck-during-downloading language servers are npm-based servers that are stuck due to networking issues when damn npm tries to access the npm registry.
Each failed network access forces this wonderful software tool to wait a bunch (alas, logs do not show how much but judging from the logs, it's a bit less than a minute) and retry each attempt, per each package, 3 times.
And these connection attempts seem to fully ignore the "regular" package fetch timeouts we have.

There had been a PR to hide such states on click: https://github.com/zed-industries/zed/pull/18305
But otherwise we seem to be out of options, so closing this issue: we need a good proposal what to do here.

**Comment by lq0910** [comment]
已收到。我会马上查看。
